In [32]:
import numpy as np
import pandas as pd

In [33]:
DATASET_AIR_QUALITY_NO2 = r"C:\Users\91930\Documents\GITHUB\ArtOfAI\dataset\air_quality\air_quality_no2_long.csv"
DATASET_AIR_QUALITY_PM25 = r"C:\Users\91930\Documents\GITHUB\ArtOfAI\dataset\air_quality\air_quality_pm25_long.csv"

In [34]:
air_quality_no2 = pd.read_csv(DATASET_AIR_QUALITY_NO2, parse_dates=True)

In [35]:
air_quality_no2.head()

city country                   date.utc location parameter  value   unit
0  Paris      FR  2019-06-21 00:00:00+00:00  FR04014       no2   20.0  µg/m³
1  Paris      FR  2019-06-20 23:00:00+00:00  FR04014       no2   21.8  µg/m³
2  Paris      FR  2019-06-20 22:00:00+00:00  FR04014       no2   26.5  µg/m³
3  Paris      FR  2019-06-20 21:00:00+00:00  FR04014       no2   24.9  µg/m³
4  Paris      FR  2019-06-20 20:00:00+00:00  FR04014       no2   21.4  µg/m³

In [36]:
# selecting the columns
air_quality_no2 = air_quality_no2[['date.utc', 'location', 'parameter', 'value']]

In [37]:
air_quality_no2

date.utc            location parameter  value
0     2019-06-21 00:00:00+00:00             FR04014       no2   20.0
1     2019-06-20 23:00:00+00:00             FR04014       no2   21.8
2     2019-06-20 22:00:00+00:00             FR04014       no2   26.5
3     2019-06-20 21:00:00+00:00             FR04014       no2   24.9
4     2019-06-20 20:00:00+00:00             FR04014       no2   21.4
...                         ...                 ...       ...    ...
2063  2019-05-07 06:00:00+00:00  London Westminster       no2   26.0
2064  2019-05-07 04:00:00+00:00  London Westminster       no2   16.0
2065  2019-05-07 03:00:00+00:00  London Westminster       no2   19.0
2066  2019-05-07 02:00:00+00:00  London Westminster       no2   19.0
2067  2019-05-07 01:00:00+00:00  London Westminster       no2   23.0

[2068 rows x 4 columns]

In [38]:
air_quality_pm25 = pd.read_csv(DATASET_AIR_QUALITY_PM25, parse_dates=True)

In [39]:
air_quality_pm25.head()

city country                   date.utc location parameter  value  \
0  Antwerpen      BE  2019-06-18 06:00:00+00:00  BETR801      pm25   18.0   
1  Antwerpen      BE  2019-06-17 08:00:00+00:00  BETR801      pm25    6.5   
2  Antwerpen      BE  2019-06-17 07:00:00+00:00  BETR801      pm25   18.5   
3  Antwerpen      BE  2019-06-17 06:00:00+00:00  BETR801      pm25   16.0   
4  Antwerpen      BE  2019-06-17 05:00:00+00:00  BETR801      pm25    7.5   

    unit  
0  µg/m³  
1  µg/m³  
2  µg/m³  
3  µg/m³  
4  µg/m³

In [40]:
air_quality_pm25 = air_quality_pm25[['date.utc', 'location', 'parameter', 'value']]

In [41]:
air_quality_pm25.head()

date.utc location parameter  value
0  2019-06-18 06:00:00+00:00  BETR801      pm25   18.0
1  2019-06-17 08:00:00+00:00  BETR801      pm25    6.5
2  2019-06-17 07:00:00+00:00  BETR801      pm25   18.5
3  2019-06-17 06:00:00+00:00  BETR801      pm25   16.0
4  2019-06-17 05:00:00+00:00  BETR801      pm25    7.5

# How to combine data from multiples tables

### Concatenating objects


![Concatenating ojbects](../../assets/pandas-combining-tables-1.png)

I want to combine the measurements of *NO<sub>2</sub>* and *PM<sub>25</sub>*, two tables with a similar structure, in a single table.

In [42]:
air_quality = pd.concat([air_quality_pm25, air_quality_no2], axis=0)

In [43]:
air_quality

date.utc            location parameter  value
0     2019-06-18 06:00:00+00:00             BETR801      pm25   18.0
1     2019-06-17 08:00:00+00:00             BETR801      pm25    6.5
2     2019-06-17 07:00:00+00:00             BETR801      pm25   18.5
3     2019-06-17 06:00:00+00:00             BETR801      pm25   16.0
4     2019-06-17 05:00:00+00:00             BETR801      pm25    7.5
...                         ...                 ...       ...    ...
2063  2019-05-07 06:00:00+00:00  London Westminster       no2   26.0
2064  2019-05-07 04:00:00+00:00  London Westminster       no2   16.0
2065  2019-05-07 03:00:00+00:00  London Westminster       no2   19.0
2066  2019-05-07 02:00:00+00:00  London Westminster       no2   19.0
2067  2019-05-07 01:00:00+00:00  London Westminster       no2   23.0

[3178 rows x 4 columns]

By default concatenation is along axis 0, so the resulting table combines the rwos of the input tables. Let's check the shape of the original and the concatenated tables to verify the operation.

In [44]:
print('Shape of the ``air_quality_pm25`` table: ', air_quality_pm25.shape)


Shape of the ``air_quality_pm25`` table:  (1110, 4)


In [45]:
print('Shape of the ``air_quality_no2`` table: ', air_quality_no2.shape)

Shape of the ``air_quality_no2`` table:  (2068, 4)


In [46]:
print('Shape of the resulting ``air_quality`` table: ', air_quality.shape)


Shape of the resulting ``air_quality`` table:  (3178, 4)


Hence, the resulting table has 3178 = 1110 + 2068 rows.



Sorting the table on the datetime information illustrates also the combination of both tables, with the `parameter` column defining the origin of the table (either `no2` from table `air_quality_no2` or `pm25` from table `air_quality_pm25`):

In [47]:
air_quality  = air_quality.sort_values('date.utc')

In [48]:
air_quality.head()

date.utc            location parameter  value
2067  2019-05-07 01:00:00+00:00  London Westminster       no2   23.0
1003  2019-05-07 01:00:00+00:00             FR04014       no2   25.0
100   2019-05-07 01:00:00+00:00             BETR801      pm25   12.5
1098  2019-05-07 01:00:00+00:00             BETR801       no2   50.5
1109  2019-05-07 01:00:00+00:00  London Westminster      pm25    8.0

In this specific example, the `parameter` column provided by the data ensures that each of the original tables can be identified. This is not always the case. The `concat` function provides a convenient colution with the `keys` argument, adding an additional (hierarchical) row index. For example:

In [49]:
air_quality_ = pd.concat([air_quality_pm25, air_quality_no2], keys=["PM25", "NO2"])

In [50]:
air_quality_

date.utc            location parameter  value
PM25 0     2019-06-18 06:00:00+00:00             BETR801      pm25   18.0
     1     2019-06-17 08:00:00+00:00             BETR801      pm25    6.5
     2     2019-06-17 07:00:00+00:00             BETR801      pm25   18.5
     3     2019-06-17 06:00:00+00:00             BETR801      pm25   16.0
     4     2019-06-17 05:00:00+00:00             BETR801      pm25    7.5
...                              ...                 ...       ...    ...
NO2  2063  2019-05-07 06:00:00+00:00  London Westminster       no2   26.0
     2064  2019-05-07 04:00:00+00:00  London Westminster       no2   16.0
     2065  2019-05-07 03:00:00+00:00  London Westminster       no2   19.0
     2066  2019-05-07 02:00:00+00:00  London Westminster       no2   19.0
     2067  2019-05-07 01:00:00+00:00  London Westminster       no2   23.0

[3178 rows x 4 columns]